fix(Project): Only users with Admin access should be allowed to edit a closed project

nutanv1contr opened the pull request:
- Only Admin should be allowed to edit a closed project.
- Added restriction, if the user is not admin and project clearing state is "Closed" then 
  user will not be allowed to edit the project.

closes : #567

maxhbr commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:
```diff
@@ -908,6 +911,7 @@ private void prepareDetailView(RenderRequest request, RenderResponse response) t
 
                 addProjectBreadcrumb(request, response, project);
                 request.setAttribute(PROJECT_OBLIGATIONS, SW360Utils.getProjectObligations(project));
+                request.setAttribute(IS_USER_ADMIN,PermissionUtils.isUserAtLeast(UserGroup.ADMIN, user) ? "Yes" : "No");
```
You should also allow `UserGroup.SW360_ADMIN` here, which is for SW360 the same as `ADMIN`. These roles only differ on the level of liferay.

nutanv1contr replied:
Thanks for the information.I have modified the configuration accordingly.

maxhbr commented on frontend/sw360-portlet/src/main/webapp/html/projects/detail.jsp:
```diff
@@ -44,7 +44,11 @@
-        <input type="button" id="edit" value="Edit" class="addButton">
+        <input type="button" id="edit" value="Edit" class="addButton"
+            <c:if test = "${(project.clearingState eq 'CLOSED') && (isUserAdmin != 'Yes')}">
++                 <c:out value="disabled='disabled'"/>
```
```suggestion
                 <c:out value="disabled='disabled'"/>
```

maxhbr replied:
Still a Problem

nutanv1contr replied:
I could not clearly understand the review comment completely here.If understood is correctly your concern is about having this "disabled" property. I have kept the feature for two reasons.
1.User will be able to immediately  know that he is not allowed to edit the project and he does not have to wait until  the response comes from backend.
2.This willl reduce the number of backend calls.
And you might already know even if user enable the "Edit" button from the browser then also user will not be able to actually edit because of the backend validation.

maxhbr replied:
This comment is only about the additional `+` at the beginning of the line.

nutanv1contr replied:
Corrected the indentation.

maxhbr commented on frontend/sw360-portlet/src/main/webapp/html/projects/detail.jsp:
```diff
@@ -44,7 +44,11 @@
+        <input type="button" id="edit" value="Edit" class="addButton"
+            <c:if test = "${(project.clearingState eq 'CLOSED') && (isUserAdmin != 'Yes')}">
++                 <c:out value="disabled='disabled'"/>
++           </c:if>
```
```suggestion
           </c:if>
```

maxhbr replied:
Still a Problem

nutanv1contr replied:
Corrected the indentation.

nutanv1contr commented:
> This change just applies the changes to the UI and not REST or the backend server (someone could still craft a change request and it will be accepted). This PR just hides the buttons. Even if someone guesses the edit link (or had it bookmarked) this PR does not stop him from editing the project.

- Now I have added backend validation which will help to validate the feature from backend server and rest end point.I have also added front end validation, which will help simplifying the user interface .

maxhbr commented on backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java:
```diff
@@ -230,6 +234,15 @@ private boolean duplicateAttachmentExist(Project project) {
 
+    private boolean updateProjectAllowed(Project project, User user) {
+        if (project.clearingState != null && project.clearingState.equals(ProjectClearingState.CLOSED)
+                && !(PermissionUtils.isUserAtLeast(UserGroup.ADMIN, user)
```
```
PermissionUtils.isUserAtLeast(UserGroup.ADMIN, user) || PermissionUtils.isUserAtLeast(UserGroup.SW360_ADMIN, user)
```
is exactly the same as only
```
PermissionUtils.isUserAtLeast(UserGroup.SW360_ADMIN, user)
```
since `UserGroup.ADMIN > UserGroup.SW360_ADMIN` it is <pre>PermissionUtils.isUser<b>AtLeast</b></pre>

nutanv1contr replied:
Thanks.Corrected the same.

maxhbr commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:
```diff
@@ -803,14 +804,24 @@ public void doView(RenderRequest request, RenderResponse response) throws IOExce
     }
 
+    private boolean isUserAdmin(User user) {
+            if ((PermissionUtils.isUserAtLeast(UserGroup.ADMIN, user)
```
```suggestion
            return PermissionUtils.isUserAtLeast(UserGroup.SW360_ADMIN, user);
```

maxhbr replied:
to many brackets.

maxhbr commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:
```diff
@@ -803,14 +804,24 @@ public void doView(RenderRequest request, RenderResponse response) throws IOExce
+        User user = UserCacheHolder.getUserFromRequest(request);
         List<Organization> organizations = UserUtils.getOrganizations(request);
         request.setAttribute(PortalConstants.ORGANIZATIONS, organizations);
+        request.setAttribute(IS_USER_ADMIN,isUserAdmin(user) ? "Yes" : "No");
```
```suggestion
        request.setAttribute(IS_USER_ADMIN,return PermissionUtils.isUserAtLeast(UserGroup.SW360_ADMIN, user) ? "Yes" : "No");
```
and remove the `isUserAdmin` function.

nutanv1contr replied:
Modified the code accordingly.

maxhbr commented on frontend/sw360-portlet/src/main/webapp/html/projects/view.jsp:
```diff
@@ -279,12 +279,18 @@
                 </core_rt:choose>
+                var editProject;
+
+		        if(row.cState == 'CLOSED' && ${isUserAdmin != 'Yes'}) {
```
formatting: we use spaces instead of tabs

maxhbr commented on libraries/lib-datahandler/src/main/thrift/sw360.thrift:
```diff
@@ -27,6 +27,7 @@ enum RequestStatus {
     FAILED_SANITY_CHECK = 4,
     DUPLICATE = 5,
     DUPLICATE_ATTACHMENT = 6,
+    UPDATE_CLOSED_PROJECT_NOT_ALLOWED = 7,
```
This is a project specific message. I would prefer a generic variant, e.g. `CLOSED_UPDATE_NOT_ALLOWED`.

nutanv1contr replied:
Good idea. Now I have made the status more generic but the error message is kept as it is. If there is any change in logic in future then the error message can be modified accordingly.

maxhbr commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:
```diff
@@ -803,8 +804,10 @@ public void doView(RenderRequest request, RenderResponse response) throws IOExce
+        User user = UserCacheHolder.getUserFromRequest(request);
         List<Organization> organizations = UserUtils.getOrganizations(request);
         request.setAttribute(PortalConstants.ORGANIZATIONS, organizations);
+        request.setAttribute(IS_USER_ADMIN,PermissionUtils.isUserAtLeast(UserGroup.SW360_ADMIN, user) ? "Yes" : "No");
```
code style:
```suggestion
        request.setAttribute(IS_USER_ADMIN, PermissionUtils.isUserAtLeast(UserGroup.SW360_ADMIN, user) ? "Yes" : "No");
```

maxhbr commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:
```diff
@@ -908,6 +911,7 @@ private void prepareDetailView(RenderRequest request, RenderResponse response) t
 
                 addProjectBreadcrumb(request, response, project);
                 request.setAttribute(PROJECT_OBLIGATIONS, SW360Utils.getProjectObligations(project));
+                request.setAttribute(IS_USER_ADMIN,PermissionUtils.isUserAtLeast(UserGroup.SW360_ADMIN, user) ? "Yes" : "No");
```
same as https://github.com/eclipse/sw360/pull/568/files#r301971841

nutanv1contr commented on frontend/sw360-portlet/src/main/webapp/html/projects/view.jsp:
```diff
@@ -279,12 +279,18 @@
+                var editProject;
+
+                if(row.cState == 'CLOSED' && ${isUserAdmin != 'Yes'}) {
+                    editProject = "<img src='<%=request.getContextPath()%>/images/edit.png' alt='Edit' title='Only Admin can edit a closed project'>";
```
The default edit.png icon is a greyed out image.
![image](https://user-images.githubusercontent.com/46559892/61197852-9abfc180-a6f4-11e9-85bd-9697a5e45e62.png)

maxhbr replied:
So, the user still does not see whether the edit icon is active or not? Why not remove it if it is inactive?

nutanv1contr replied:
Now user gets a message on mouse hover of the icon that "Only admin can edit a closed project".If we remove the icon then user may get confused about the reason why the button is not there.

maxhbr replied:
mouse-over is not enough. I want to see whether I am able to edit something. Maybe change the icon to something with 40% opacity.

nutanv1contr replied:
Made the necessary changes as suggested.Added the opacity level for the disabled edit button. Now there will be a clear discrimination of the disabled edit button.
![image](https://user-images.githubusercontent.com/46559892/61221700-a92ecd00-a736-11e9-94d0-c5207d817746.png)

mcjaeger commented:
@maxhbr comments addressed?

maxhbr commented on backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java:
```diff
@@ -230,6 +234,14 @@ private boolean duplicateAttachmentExist(Project project) {
     }
 
+    private boolean updateProjectAllowed(Project project, User user) {
+        if (project.clearingState != null && project.clearingState.equals(ProjectClearingState.CLOSED)
```
minor simplification:
```suggestion
        return ! (project.clearingState != null && project.clearingState.equals(ProjectClearingState.CLOSED) && !PermissionUtils.isUserAtLeast(UserGroup.SW360_ADMIN, user));
```

maxhbr commented on frontend/sw360-portlet/src/main/webapp/html/projects/detail.jsp:
```diff
@@ -44,7 +44,11 @@
-        <input type="button" id="edit" value="Edit" class="addButton">
+        <input type="button" id="edit" value="Edit" class="addButton"
+            <c:if test = "${(project.clearingState eq 'CLOSED') && (isUserAdmin != 'Yes')}">
+                <c:out value="disabled='disabled'"/>
```
Minor comment: escaping via `<c:out` is unnecessary:
```suggestion
                disabled="disabled"
```

maxhbr replied:
The `<c:out` tag is usually used to access fields in java objects, e.g. something like `<c:out value="customer.address.street"/>`. (see https://www.tutorialspoint.com/jsp/jstl_core_out_tag). This means this code interprets `disabled='disabled'` as an expression.

This should be removed.

nutanv1contr replied:
Done

maxhbr commented on rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java:
```diff
@@ -75,7 +75,9 @@ public Project createProject(Project project, User sw360User) throws TException
         RequestStatus requestStatus = sw360ProjectClient.updateProject(project, sw360User);
-        if (requestStatus != RequestStatus.SUCCESS) {
+        if (requestStatus != RequestStatus.CLOSED_UPDATE_NOT_ALLOWED) {
+            throw new RuntimeException("User cannot modify a closed project");
```
Minor comment: maybe also log which project and which user?
